### DISCUSSION TITLE
Synology package (SPK) that contains hkknx

### DISCUSSION THREAD

**Discussion by nanosonde** [brutella/hkknx-public#122] [Ideas]
An SPK file seems to be a TAR file only which contains some basic info and scripts to get things up and running:
Example: https://github.com/openhab/openhab-syno-spk

The alternative way would be a docker image of course.
Synology SPK can offer ARM and/or x86 architecure.

As hkknx is a self-contained GO-binary it should be quite easy IMHO.

**Comment by nanosonde** [comment]
There is also a blog post here which describes the steps required: https://neerajcodes.wordpress.com/2017/03/08/creating-a-package-for-synology-nas-device/

It uses the official Synology package toolkit: https://github.com/SynologyOpenSource/pkgscripts-ng

Official docs here: https://www.synology.com/de-de/support/developer#tool

**Comment by dj95** [comment]
For a docker container, I prepared a Dockerfile and docker-compose.yml (in case you are also interested in this solution as you've picked it up in the description). Please bear in mind, that the container must be run with `network_mode=host` (respective `--net=host` on the command line) in order to enable hkknx for communicating with the ip interface.

**Dockerfile**

```Dockerfile
#
# Step 1 - Download and extract hkknx
#

# use alpine as a build environment
FROM alpine as downloader

# define the hkknx version as build argument to create a dynamic dockerfile
ARG HKKNX_VERSION=2.0.0

# define the hkknx architecture as build argument to create a dockerfile for arm
# and x86
ARG HKKNX_ARCH=linux_amd64

# install wget, download hkknx and unpack it
RUN apk add --no-cache wget \
 && mkdir /app \
 && cd /app \
 && wget https://github.com/brutella/hkknx-public/releases/download/${HKKNX_VERSION}/hkknx-${HKKNX_VERSION}_${HKKNX_ARCH}.tar.gz \
 && tar xvzf hkknx-${HKKNX_VERSION}_${HKKNX_ARCH}.tar.gz

#
# Step 2 - Download ca certificates for external communication
#

# use alpine again to get the certificates via apk
FROM alpine AS ca-certificates

# install the ca certificates
RUN apk add --no-cache ca-certificates


#
# Step 3 - Bundle the downloaded dependencies and build the runtime container
#

# start with an empty image
FROM scratch

# copy the hkknx binary
COPY --from=downloader /app/hkknx /hkknx

# copy the ca certificates
COPY --from=ca-certificates /etc/ssl /etc/ssl

# define a default volume for the default database
VOLUME /database

# entrypoint will be hkknx
ENTRYPOINT ["/hkknx"]

# define standard commandline arguments
CMD ["-db", "/database"]
```

**docker-compose.yml**

```yaml
version: '3.5'
services:
  hkknx:
    build:
      context: '.'
      args:
        HKKNX_VERSION: '2.0.0'
        HKKNX_ARCH: 'linux_amd64'
    image: 'hkknx'
    network_mode: 'host'
```
